ui/components/app/confirm/signature-message/signature-message.tsx

```diff
@@ -0,0 +1,56 @@
+import React, { memo, useMemo } from 'react';
+import { useSelector } from 'react-redux';
+
+import { MESSAGE_TYPE } from '../../../../../shared/constants/app';
+import {
+  BackgroundColor,
+  BorderRadius,
+} from '../../../../helpers/constants/design-system';
+import { hexToText } from '../../../../helpers/utils/util';
+import { useI18nContext } from '../../../../hooks/useI18nContext';
+import { currentConfirmationSelector } from '../../../../selectors';
+import { Box } from '../../../component-library';
+import { ConfirmInfo, ConfirmInfoRowType } from '../info/info';
+
+const SignatureMessage: React.FC = memo(() => {
+  const t = useI18nContext();
+  const currentConfirmation = useSelector(currentConfirmationSelector);
+
+  const args = useMemo(() => {
```

Comment:
Why is this architecture preferable than just using plain JSX? It seems like this would be harder to maintain without much added benefit. Unless I am missing something?

Reply:
This gives us more flexibility as these components will be reused for different types of confirmations.
And I actually do not see any maintenance overhead in keeping this logic in / out of jsx.

Reply:
Could you elaborate on how this system "gives us more flexibility"? I'm having some trouble following how this config maps to components, but it seems like this would be simpler to understand using more standard techniques.

Reply:
@Gudahtt : we have multiple confirmation pages with different combination of this information displayed. I see this new info section more as tabular display of this data.

![Screenshot 2024-02-09 at 4 32 53 PM](https://github.com/MetaMask/metamask-extension/assets/2182307/4ada72d0-04d4-4d0c-a5d9-30a9957d90ad)

And using array to pass data to `ConfirmInfo` components gives us flexibility. But wherever we have fixed set of data like this example of message it will be better to use the jsx version.

Reply:
Ref: https://github.com/MetaMask/metamask-extension/pull/22763

Reply:
@jpuri Apologies, but I still don't see how this approach gives us any extra flexibility.

I have significant worries that this adds unnecessary complexity to these screens, that already will be complex due to their nature. I don't currently see any benefit to introducing an additional pattern for these screens that differs from everything else in the extension codebase.

I think it would be great if we could get some more opinions on this, I just want to make sure that we're introducing a pattern that may become problematic later on, especially this early in the redesign.

Perhaps @brad-decker would be a good person to give some feedback on this? Or would have ideas who else should be brought in to give their opinion on this.

Regardless, I would very much be opposed to this pattern unless I can be convinced of its usefulness.

Reply:
Hopefully approaching this in as least opinionated way as possible:

Composing props in this format of configurations, wherein each entry in the array informs what type of content is being rendered in the row (and which component should render it) is only moving the complexity to a different level. 

The pure JSX way of handling this with type safety would be to have a "RowSwitch" component that checks the `type` key and then returns the proper component. However that doesn't negate from my perspective the need for an array of unions being supported? You need to pass the type of value and the value. The only difference would be that you'd have a TSX file that looks something like this psuedocode:

```tsx

export function RowSwitch(row: RowType) {
  switch(row.type) {
    case Types.Text:
       return <Text value={row.value} />;
     case Types.Other:
       return <Other value={row.value} />;
     }
};
```

Which would then allow you to loop over the rows of the array and render <RowSwitch /> in this file instead of doing so in the ConfirmInfo component. So your signature-message file would look more like:

```tsx
  return (
    <Box
      backgroundColor={BackgroundColor.backgroundDefault}
      borderRadius={BorderRadius.MD}
      padding={2}
      marginBottom={4}
    >
      {rows.map((row) => <ConfirmInfoRow key={row.label} row={row} />}
    </Box>
  );
  ```
  
  with ConfirmInfoRow having the RowSwitch component in it. 
  
  I'm not entirely sure whats being objected to here @FrederikBolding if its the pattern of passing an array then I think you're going to have to deal with that array somewhere. I agree slightly that dealing with it in the file you're working on is better than letting another file render it, but its a very slight agreement. If the concern is in **how** we are choosing what renders then I think its a matter of preference.

Reply:
> I agree slightly that dealing with it in the file you're working on is better than letting another file render it, but its a very slight agreement.

+1

Reply:
>  if its the pattern of passing an array then I think you're going to have to deal with that array somewhere.

@brad-decker Can you elaborate on this? I'm not sure I follow how this is a necessity.

From my perspective building these screen using plain JSX e.g:

```tsx
    <Box
      backgroundColor={BackgroundColor.backgroundDefault}
      borderRadius={BorderRadius.MD}
      padding={2}
      marginBottom={4}
    >
      <ConfirmInfoRow label="Message">
        <ConfirmInfoRowText text={hexToText(currentConfirmation.msgParams?.data)} />
      </ConfirmInfoRow>
      {someCondition && <ConfirmInfoRow label="Foo bar">
        <ConfirmInfoRowText text="Some text" />
      </ConfirmInfoRow>}
    </Box>
```

is much more readable and probably also more performant than:
```tsx

  const args = useMemo(() => {
    if (
      !currentConfirmation ||
      currentConfirmation.type !== MESSAGE_TYPE.PERSONAL_SIGN ||
      !currentConfirmation.msgParams?.data
    ) {
      return null;
    }
    return {
      rowConfigs: [
        {
          label: t('message'),
          type: ConfirmInfoRowType.Text,
          rowProps: {
            text: hexToText(currentConfirmation.msgParams?.data),
          },
        },
        {
          label: 'foo',
          type: ConfirmInfoRowType.Text,
          rowProps: {
            text: 'bar'
          },
        },
      ],
    };
  }, [currentConfirmation]);

return (
    <Box
      backgroundColor={BackgroundColor.backgroundDefault}
      borderRadius={BorderRadius.MD}
      padding={2}
      marginBottom={4}
    >
      <ConfirmInfo {...args} />
    </Box>
  );
```

I just don't see the reasoning or what we gain by adding an additional abstraction on top of this? Jyoti argues that it is more flexible, but I have yet to see an example of how?

Reply:
> >if its the pattern of passing an array then I think you're going to have to deal with that array somewhere.
>@brad-decker Can you elaborate on this? I'm not sure I follow how this is a necessity.

@FrederikBolding I suppose I was laboring under the illusion that these screens would, at some point in the future, be able to interface with snaps that modify the screen's content. If we control this screen in its entirety, and we do not foresee the need to allow inroads into sources other than our code files to make changes here, then yes by all means use JSX syntax. My context for what the confirmations team is trying to achieve is dated.

However, from the standpoint of performance, rendering components from an array of data is a typical pattern in React, and the performance of doing it is subject to the rules of React (mapping over an array of transactions as we do in the activity list, for example). You can make unoptimized JSX, and you can make unoptimized rendering of arrays, and you can do both -- but assuming we are following best practices, doing what is being done in this PR is not inherently slower. The array is finite, the configuration is limited, and the scope is small.

From a stylistic and readability, I strongly prefer we use JSX if we do not need to modify the data structure programmatically. Meaning if we do not now know, nor do we foresee, of a use case for providing flexibility in the structure then I agree 100% with what @FrederikBolding has said on this subject.

Reply:
IMHO, jsx is not the best place for conditional, plain JS or TS code is better at that. These components will take care of different confirmation type with differing data to be displayed, this nested conditional in JSX does not look like a good idea to me.

I am not against either of the approaches, I also agree its more of preference, but I do not see a downside of current approach either.

Reply:
> JSX is syntactic sugar

Exactly, "syntactic sugar" making the code more readable (at least to my eyes, fwiw)

Reply:
_As someone not as heavily involved with this code as everyone else is in this thread, reading (conditional) JSX seems much easier to me than what is proposed in the 2nd example_